libavcodec/libaribcaption.c

```diff
@@ -0,0 +1,1141 @@
+                    m = div_a - n * y0;
+                    idx0 = image->stride * y0 + w * 4;
+                    idx1 = image->stride * y1 + w * 4;
+                } else {
+                    int div_a, x0, x1;
+                    div_a = w * ctx->frame_width;
+                    n = ctx->bitmap_plane_width;
+                    x0 = div_a / n;
+                    x1 = FFMIN(x0 + 1, image->width - 1);
+                    m = div_a - n * x0;
+                    idx0 = image->stride * h + x0 * 4;
+                    idx1 = image->stride * h + x1 * 4;
+                }
+                r = (image->bitmap[idx0++] * (n - m) + image->bitmap[idx1++] * m) / n;
+                g = (image->bitmap[idx0++] * (n - m) + image->bitmap[idx1++] * m) / n;
+                b = (image->bitmap[idx0++] * (n - m) + image->bitmap[idx1++] * m) / n;
+                a = (image->bitmap[idx0++] * (n - m) + image->bitmap[idx1++] * m) / n;
+                rect->data[0][dst_idx++] = clut_pick_or_set(ctx, r, g, b, a);
+            }
+        }
+        rect->data[1] = av_memdup(ctx->clut, AVPALETTE_SIZE);
+        if (!rect->data[1]) {
+            ret = AVERROR(ENOMEM);
+            goto fail;
+        }
+
+        rect->x = image->dst_x * ctx->bitmap_plane_width / ctx->frame_width;
+        rect->y = image->dst_y * ctx->bitmap_plane_height / ctx->frame_height;
+        rect->type = SUBTITLE_BITMAP;
+        rect->linesize[0] = rect->w;
+        rect->nb_colors = 256;
+
+        ff_dlog(ctx, "BITMAP subtitle%s (%d,%d) %dx%d -> (%d,%d) %dx%d [%d]: %d colors\n",
+                (ctx->caption.regions[rect_idx].is_ruby) ? " (ruby)" : "",
+                image->dst_x, image->dst_y, image->width, image->height,
+                rect->x, rect->y, rect->w, rect->h,
+                rect_idx, ctx->clut_idx);
+        if (ctx->clut_overflow)
+            av_log(ctx, AV_LOG_WARNING, "CLUT overflow (%d).\n", ctx->clut_overflow);
+    }
+    sub->num_rects = rect_idx;
+
+    return rect_idx;
+
+fail:
+    if (sub->rects) {
+        for (int i = 0; i < ctx->caption.region_count; i++) {
+            if (sub->rects[i]) {
+                av_freep(&sub->rects[i]->data[0]);
+                av_freep(&sub->rects[i]->data[1]);
+                av_freep(&sub->rects[i]);
+            }
+        }
+        av_freep(&sub->rects);
+    }
+    sub->num_rects = 0;
+
+    return ret;
+}
+
+static int set_ass_header(ARIBCaptionContext *ctx)
+{
+    AVCodecContext *avctx = ctx->avctx;
+    int outline, shadow;
+    const char *font_name;
+    const char *fonts = ctx->font;
+
+    if (ctx->border_style == 4) {
+        outline = 0;
+        shadow = 4;
+    } else {
+        outline = 1;
+        shadow = 0;
+    }
+    if (ctx->force_stroke_text)
+        outline = (int)(ctx->stroke_width * 4.0 / 3.0);
+
+    if (fonts && *fonts)
+        font_name = av_get_token(&fonts, ",");
+    else
+        font_name = av_strdup(DEFAULT_FONT_FAMILY);
```

Comment:
Why not just make `sans-serif` the default for the AVOption?

Reply:
The `libaribcaption` library has it's default font for bitmap depending on operating system. I would like to follow it if `font` option is not specified.
On the other hand, we have to default font for ASS subtitle.

Reply:
This isn't using a default from libaribcaption, though? It's just using `DEFAULT_FONT_FAMILY`, which is defined as `sans-serif`.

Reply:
`libaribcaption` renders (with default font) when `sub_type` is `bitmap` only.
When `sub_type` is `ass`, `libaribcaption` parses ARIB caption, but does not render nor encode to ASS text.
The `DEFAULT_FONT_FAMILY` is set to ASS header only, but most players such as `mpv` set its own ASS header and specified font is ignored.

Reply:
Ahh, maybe name the macro `ASS_DEFAULT_FONT` or the like for clarity?

Reply:
`ASS_DEFAULT_FONT` is already defined in `libavcodec/ass.h` but is `"Arial"`.
The font is not suitable for ARIB caption.
Which is your preference, redifine with same macro name or define another new (but ASS related) macro name?

Reply:
Hmm, maybe `DEFAULT_FONT_ASS`? And add a comment indicating that it's only used for ASS, because the bitmap renderer has its own internal default.